### PR TITLE
Samuel npm/web fixes

### PR DESCRIPTION
**Pull request by engineersamuel**
One of the core issues with the npm build is the exr-wrap.wasm is loaded within the exr-wrap.js as an XMLHttpRequest which uses the location.href which, in your own React app, resolves to your own React app. That is unavoidable, the way it's currently written.  The best solution for this was to simplify the build so everything is built through webpack which will then dynamically load the wasm as a function.

- Unified the npm and web builds, now there is only 1 that works both in npm and the web.
- Updated the package.json dependencies to more recent versions.
- Created an examples directory that is persisted in git, the build/ dir now can be safely removed and re-built without worrying about examples.  When `npm run examples` the build/* contents are copied to the examples dir.
- Documented how to npm-sync to do local dev on this, and also how to make it work with Create React App
- Enhanced the image loading to allow for a data uri and created a test example demonstrating that.
- Minor other refactorings.

### CODE REVIEW

**Comment by cbreak-black** [comment]
Thanks! We'll perform some internal tests to see how this affects our internal use of jeri, I hope we can give detailed feedback soon

**Comment by cbreak-black** [comment]
Our comments (@cbreak-black and @maxgrosse ):
- I had to restore the example EXR Images, those seem to have been lost in the move from `build_web` to `build`
- I had to restore `"examples": "serve build"` in `package.json`, to run our example server / "test"
- Then I was able to run the example successfully :D
- please remove the unneeded junk files `src/*.swp`
- The files from `exr-wrap` are generated, you should not change them. In particular, your changes to `src/exr-wrap/exr-wrap.js` would be overwritten the next time I update the wrapper with a fresh build.
- Why did you replace the `let` with `var` in the file `src/utils/exr-parser.worker.js`? I read that `let` is generally preferred, since the scoping of `let` is much tighter, and less error prone. Similarly, `const` was also replaced by `var`.
- What is the reasoning behind `exrImage["delete"]();`? The commit message says that this is supposed to help with `npm i`, but I don't see how that's related.

**Comment by maxgrosse** [comment]
Thanks a lot @engineersamuel this is great and really appreciated! I think we are happy to merge once the above points have been addressed!

**Comment by engineersamuel** [comment]
@cbreak-black Thank you for the feedback!  I'll address the points:

- [x] Copy that on the exr image, I'll re-evaluate the examples, I had assumed anything in the build_* dirs were deletable, so I'll need to put those in the examples dir
- [x] I'll update the package json for serve build in the next commit
- [x] with *.swp my mistake, I use vim and looks like I didn't update the .gitignore
- So the exr-wrap is going to be an issue.  I had to update it to allow for wasm to be loaded as a function with Webpack.  This is the only way to unify the npm and web builds.  It's possible newer versions of whatever libs are updating the wrapper handle this?  If they don't, my changes will need to be manually merged in each time. But from my understanding generating that wasm build is rare right?
- [x] On the exr-parser.worker.js changes, they were inadvertent, I have a todo item to revert those changes as they were from internal testing that I had committed unnecessarily.  Look for updates in the next commit.
- As for the exrImage["delete"] I did not directly make that change, I'll evaluate it.

To recap, I'll push updates for all points except the changes to exr-wrap.js, see my point above, we'll have to figure that out.

**Comment by engineersamuel** [comment]
@maxgrosse Thank you, happy to contribute!

**Comment by cbreak-black** [comment]
>     * So the exr-wrap is going to be an issue.  I had to update it to allow for wasm to be loaded as a function with Webpack.  This is the only way to unify the npm and web builds.  It's possible newer versions of whatever libs are updating the wrapper handle this?  If they don't, my changes will need to be manually merged in each time. But from my understanding generating that wasm build is rare right?

Yes, this is very rare. The `.wasm` file, and the associated loader stub were generated by emscripten. I've not run it in years, so I'm quite sure everything in emscripten changed since then.

> To recap, I'll push updates for all points except the changes to exr-wrap.js, see my point above, we'll have to figure that out.

Yeah :/ We don't have any immediate need to update the wrapper either, but were a bit concerned about the potential in the future.

File: src/exr-wrap/exr-wrap.js [hunk 1]
```diff
@@ -1,24 +1,3359 @@
...
+            return false;
+          }
+
+          function integrateWasmJS() {
+              var wasmTextFile = "exr-wrap.wast";
+              // Note that we always want to require here and not let the fallback be to load via XMLHttpRequest if it's text.
+              // The issue here is that in a parent npm project that requires jeri, if this exr-wrap.js and below requires isn't pre-processed
+              // with webpack, then the parent project doesn't know what to do with it, or we require the parent project to deal with it,
+              // which really isn't an option, therefore require the wasm and handle it here.
+              var wasmBinaryFile = require("./exr-wrap.wasm");
```
Comment (engineersamuel) [inline src/exr-wrap/exr-wrap.js]: @cbreak-black Tagging you here, it's unfortunate on the diff.  I unminified it so I could fix the webpack pieces, but then it causes the diff to appear that I changed everything, and I really didn't.  This is one line I changed, that explains why.  Without a require, webpack won't handle this piece, when webpack doesn't handle it, it causes an XMLHttpRequest both in the browser and npm.  The require let's webpack nicely handle it with the wasm-loader, which is really what we want to happen here.

File: src/exr-wrap/exr-wrap.js [hunk 2]
```diff
@@ -1,24 +1,3359 @@
...
+              var asmjsCodeFile = "exr-wrap.temp.asm.js";
+              if (!isDataURI(wasmTextFile)) {
+                  wasmTextFile = locateFile(wasmTextFile)
+              }
+              if (!isDataURI(wasmBinaryFile) && typeof wasmBinaryFile !== "function") {
+                  wasmBinaryFile = locateFile(wasmBinaryFile)
+              }
```
Comment (engineersamuel) [inline src/exr-wrap/exr-wrap.js]: Here I added the `!== "function"` so the locateFile wouldn't be called to trigger the XMLHttpRequest

File: src/exr-wrap/exr-wrap.js [hunk 3]
```diff
@@ -1,24 +1,3359 @@
...
+                      getBinaryPromise().then((function(binary) {
+                          console.log(`binary: ${binary}`);
+                          return WebAssembly.instantiate(binary, info)
+                      })).then(receiver).catch((function(reason) {
+                          err("failed to asynchronously prepare wasm: " + reason);
+                          abort(reason)
+                      }))
+                  }
+                  
+                  if (typeof wasmBinaryFile === "function") {
```
Comment (engineersamuel) [inline src/exr-wrap/exr-wrap.js]: @cbreak-black  this is one of the few other pieces I added, just to handle when the wasmBinaryFile is a function loaded by webpack.

File: src/exr-wrap/exr-wrap.js (hunk 1 above)
Comment (maxgrosse) [reply src/exr-wrap/exr-wrap.js]: great, maybe we can check if emscripten has an option to have this done itself nowadays, or there might be a way to fix this otherwise on the generator side

Comment (engineersamuel) [reply src/exr-wrap/exr-wrap.js]: I figured out a solution I'll resolve this and tag you

File: src/utils/exr-parser.worker.js [hunk 4]
```diff
@@ -17,35 +17,34 @@
...
-		const job = queuedJobs.shift();
-		if (job) {
-			handleJob(job);
-		}
-	}
+// The wasmBinary will be read in as a function to be instantiated in the exr-wrap.js
+const wasmBinary = require('../exr-wrap/exr-wrap.wasm');
+const EXR = require("../exr-wrap/exr-wrap.js");
+EXR({
+  instantiateWasm: (info, receiveInstance) => wasmBinary(info).then(instance => receiveInstance(instance.instance))
```
Comment (engineersamuel) [inline src/utils/exr-parser.worker.js:37]: @cbreak-black After your comments yesterday it helped me take a step back, I realized that surely there is a way to do this without modifying the exr-wrap.js which shouldn't be modified, as you pointed out.

And yes there is.  I hacked it into the exr-wrap.js, and if I hacked it in, there should be a way to call that exr-wrap.js with the wasm, and there is, here is the line to invoke it.  To my knowledge this isn't documented anywhere I could find, can only find it from beautifying the exr-wrap.js and looking how it actually is instantiating the wasm, then mimicking that on the EXR module itself, works beautifully.  

So the reason this is crucial, this is what allows the unified build to work without dealing with handling the wasm externally.  Previously if a project wanted to use jeri, they have to host wasm and have it load from `location.origin` which can be problematic.  Now, it's embedded within the jeri.js itself and dynamically loaded as a function with no external XMLHttpRequest calls.

**Comment by engineersamuel** [comment]
@cbreak-black Ok this is ready for a final review.  See my comment I tagged you in, and I updated the first comment with an overview of work done.  Also I've tested:

- [x] In the Visualizer which does `npm i -D jeri`
- [x] In the browser examples
- [x] In an AML notebook

And Jeri loads and renders both exr and regular images as expected in all three scenarios.  See the updated readme and examples structure as well.

File: src/utils/exr-parser.worker.js (hunk 4 above)
Comment (cbreak-black) [reply src/utils/exr-parser.worker.js:37]: Cool! I'm not sure how fragile the "ABI" of the wasm compiled part is, but this already looks like a far superior solution.

File: README.md [hunk 5]
```diff
@@ -18,12 +18,12 @@ For users experienced with web development that build their projects with [webpa
 To use JERI Viewer in a web application that is built with webpack, run
 
 ```bash
-npm install --save react
-npm install --save react-dom
-npm install --save jeri
+npm i -D react
```
Comment (maxgrosse) [inline README.md:21]: the typical user might want `-S` instead, iirc `-D` is saving it as development dependency?

**Comment by maxgrosse** [comment]
I think @cbreak-black confirmed it all runs and works, in which case I'd be fine with merging it! (dunno about the `-D` vs `-S` above though, but that just affects the documentation so..)

File: README.md (hunk 5 above)
Comment (engineersamuel) [reply README.md:21]: As far as I can tell, jeri will only ever be used in a browser context, even when installed with npm.  Therefore in a web project, like React, which you'd `npm i jeri` you'd want it in the `devDepedencies` considering that it's used to build the final production artifacts to be used in the browser, but not technically used to run in production (node).  Therefore I think `-D` is most appropriate here.

Comment (maxgrosse) [reply README.md:21]: oh, right. Absolutely agree! Thanks!

**Comment by engineersamuel** [comment]
@cbreak-black @maxgrosse Awesome! What's your plan with master, so this would merge into develop, would there be a plan to merge develop to master eventually?

**Comment by maxgrosse** [comment]
that's a good question, I think @cbreak-black and me should carve out some time in the near future to discuss this. I don't really see any reason why not taking this opportunity to merge into master as well, but I'd need to check who depends on it etc.

**Comment by cbreak-black** [comment]
> @cbreak-black @maxgrosse Awesome! What's your plan with master, so this would merge into develop, would there be a plan to merge develop to master eventually?

Yes, we've been planing to release a new master for years, but there wasn't much interest, and then suddenly there was too much interest for it :) I think releasing a new master makes total sense, once the changes have stabilized.

Next step for me will be to try out this merge in our internal codebase, and possibly make changes required for it.
